plugins/catalog/src/components/EntityRelationWarning/EntityRelationWarning.tsx

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Entity, stringifyEntityRef } from '@backstage/catalog-model';
+import {
+  CatalogApi,
+  catalogApiRef,
+  useEntity,
+} from '@backstage/plugin-catalog-react';
+import { Alert } from '@material-ui/lab';
+import React from 'react';
+import useAsync from 'react-use/lib/useAsync';
+import { Box } from '@material-ui/core';
+import { ResponseErrorPanel } from '@backstage/core-components';
+import { useApi, ApiHolder } from '@backstage/core-plugin-api';
+
+async function getRelationWarnings(entity: Entity, catalogApi: CatalogApi) {
+  const entityRefRelations = entity.relations?.map(
+    relation => relation.targetRef,
+  );
+  if (!entityRefRelations || entityRefRelations?.length < 1) {
+    return [];
+  }
+
+  const relatedEntities = await catalogApi
+    .getEntitiesByRefs({ entityRefs: entityRefRelations })
+    .then(entities => entities.items.filter(_ => _ !== undefined) as Entity[]);
+
+  return entityRefRelations.filter(
```

Comment:
This calls stringifyEntityRef N*M times, which gets pretty expensive for very large numbers of relations.

Note that the response from the catalog has nulls _in the same array index_ as the red you asked for. You don't need to filter and search. You just need to iterate through the _original input array_ of refs, and check for nulls in the corresponding response slot.

Reply:
Good point regarding the index. I hope the order/index will stay stable in the future for `getEntitiesByRefs`.

Reply:
Yeah that is definitely explicitly in the contract!